File: shared/protocol.ts
// Shared WebSocket protocol between web client and Rust server.
// Keep in sync with src-tauri/src/protocol.rs

export type MouseButton = "left" | "right" | "middle";

export type MediaAction =
  | "play_pause"
  | "next"
  | "prev"
  | "vol_up"
  | "vol_down"
  | "mute";

export type SystemAction =
  | "slide_next"
  | "slide_prev"
  | "lock"
  | "sleep"
  | "shutdown";

// Special (non-text) keys the keyboard panel can send.
export type SpecialKey =
  | "enter"
  | "escape"
  | "tab"
  | "backspace"
  | "delete"
  | "space"
  | "up"
  | "down"
  | "left"
  | "right"
  | "home"
  | "end"
  | "pageup"
  | "pagedown"
  | "win"
  | "copy"
  | "paste";

export type ClientMessage =
  | { type: "auth"; token: string }
  | { type: "mouse_move"; dx: number; dy: number }
  | { type: "mouse_abs"; x: number; y: number }
  | { type: "mouse_click"; button: MouseButton }
  | { type: "mouse_double"; button: MouseButton }
  | { type: "mouse_down"; button: MouseButton }
  | { type: "mouse_up"; button: MouseButton }
  | { type: "mouse_scroll"; dy: number }
  | { type: "key"; key: SpecialKey }
  | { type: "text"; text: string }
  | { type: "media"; action: MediaAction }
  | { type: "system"; action: SystemAction }
  | { type: "ping" };

export type ServerMessage =
  | { type: "status"; ok: boolean }
  | { type: "authed" }
  | { type: "error"; msg: string }
  | { type: "pong" };

export interface PairResponse {
  token: string;
}

File: src-tauri/src/protocol.rs
use serde::{Deserialize, Serialize};

// Mirrors shared/protocol.ts — keep in sync.

#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum MouseButton {
    Left,
    Right,
    Middle,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum MediaAction {
    PlayPause,
    Next,
    Prev,
    VolUp,
    VolDown,
    Mute,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum SystemAction {
    SlideNext,
    SlidePrev,
    Lock,
    Sleep,
    Shutdown,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum SpecialKey {
    Enter,
    Escape,
    Tab,
    Backspace,
    Delete,
    Space,
    Up,
    Down,
    Left,
    Right,
    Home,
    End,
    PageUp,
    PageDown,
    Win,
    Copy,
    Paste,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMessage {
    Auth { token: String },
    MouseMove { dx: f64, dy: f64 },
    MouseAbs { x: f64, y: f64 },
    MouseClick { button: MouseButton },
    MouseDouble { button: MouseButton },
    MouseDown { button: MouseButton },
    MouseUp { button: MouseButton },
    MouseScroll { dy: f64 },
    Key { key: SpecialKey },
    Text { text: String },
    Media { action: MediaAction },
    System { action: SystemAction },
    Ping,
}

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
#[allow(dead_code)] // Status reserved for future server->client push
pub enum ServerMessage {
    Status { ok: bool },
    Authed,
    Error { msg: String },
    Pong,
}
